driver/clock_control: Clean up stm32h7 and stm32u5 drivers

erwango opened the pull request:
Attempt to clean up STM32 clock_control driver, starting on H7 and U5 series.
Main target is to clean up `clock_stm32_ll_common.c` which is used by all the other STM32 series,
but I first want to validate ideas on these 2 first ones.

In the end, there are not much changes, aside from:
- factorization when possible
- use of IS_ENABLED when possible to ease readability

Changes were tested thanks to new dedicated tests available in `tests/drivers/clock_control/stm32_clock_configuration`.

Based on:
- ~~https://github.com/zephyrproject-rtos/zephyr/pull/41310~~ (merged)
- ~~https://github.com/zephyrproject-rtos/zephyr/pull/41391~~ (merged)

EDIT (01/18):
Following changes were made on top of existing proposal. Reason not to split this in another PR is that I prefer to group breaking changes in one big PR and then be done with it. I think this is also preferable for driver users.
- Reworked various clocks sources configuration and enabling. They're now configured independently and in logical order (fixed clocks, then PLLs) at startup based on device tree status. Then they are used if required when configuring sysclock
- Reworked some STM32_CLOCK dt symbols in order to simplify usage in clock_control drivers.

The rationale behind this new changes is to prepare introduction of alternate / optional clock sources (cf #41650), which requires to be able to enable each source clock independently from each others.

erwango committed:
include/drivers/clock_control: stm32: Remove some conditions

Remove conditions around some definitions since these symbols
and the node itself are mandatory.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
include/driver/clock_control: stm32: Rearrange definitions

Group definitions by related node.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
include/drivers/clock_control: stm32: Make few macros IS_ENABLED-able

Tweak some macros definitions so that they can be used in
IS_ENABLED utility macro.
Finality is to rework STM32 clock_control driver to a more
readable format.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/clock_control: stm32h7: Make driver more readable

Re-arrange code using benefits of IS_ENABLED.
Change some #if to #ifdef when possible.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/clock_control: stm32h7: Complementary readability changes

- Factorize elementary clocks setup code.
- Put conditional logic on CONFIG_CPU_CORTEX_M4 outside init function

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/clock_control: stm32u5: Make driver more readable

Use benefits of IS_ENABLED.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

ABOSTM reviewed (changes requested):
Few inline comments, otherwise, LGTM

erwango committed:
drivers/clock_control stm32h7: Move RCC HW semaphore unlock

PLL3 setting should also be protected CFG_HW_RCC_SEMID.
Move semaphore unlock after we're done with PLL3.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@ABOSTM Thanks for the review. Comments updated.

ABOSTM reviewed (dismissed):
LGTM

FRASTM commented on drivers/clock_control/clock_stm32_ll_u5.c:
```diff
@@ -51,29 +51,29 @@ void config_enable_default_clocks(void)
-	while (READ_BIT(RCC->BDCR, RCC_BDCR_LSESYSRDY) == 0U) {
-	}
+		/* Enable LSESYS additionnally */
+		SET_BIT(RCC->BDCR, RCC_BDCR_LSESYSEN);
```
this is `LL_RCC_LSE_EnablePropagation()` in stm32U5

FRASTM commented on drivers/clock_control/clock_stm32_ll_u5.c:
```diff
@@ -51,29 +51,29 @@ void config_enable_default_clocks(void)
+		/* Enable LSESYS additionnally */
+		SET_BIT(RCC->BDCR, RCC_BDCR_LSESYSEN);
+		/* Wait till LSESYS is ready */
+		while (READ_BIT(RCC->BDCR, RCC_BDCR_LSESYSRDY) == 0U) {
```
this is `LL_RCC_LSESYS_IsReady()` in stm32U5

FRASTM commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -510,102 +506,97 @@ static struct clock_control_driver_api stm32_clock_control_api = {
-	/* Enable HSI oscillator */
-	LL_RCC_HSI_Enable();
-	while (LL_RCC_HSI_IsReady() != 1) {
+	LL_RCC_PLL3_Enable();
```
could we have 	`if (IS_ENABLED(STM32_HSI_ENABLE)) {`

FRASTM commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -510,102 +506,97 @@ static struct clock_control_driver_api stm32_clock_control_api = {
+	r = get_vco_input_range(STM32_PLL3_M_DIVISOR, &vco_input_range);
+	if (r < 0) {
+		return r;
+	}
```
Is the `stm32_clock_control_init` function returning 0 ?
In the clock_stm32_ll_common.c we read:
```
/**
 * @brief Initialize clocks for the stm32
 *
 * This routine is called to enable and configure the clocks and PLL
 * of the soc on the board. It depends on the board definition.
 * This function is called on the startup and also to restore the config
 * when exiting for low power mode.
 *
 * @param dev clock device struct
 *
 * @return 0
 */
```

erwango replied:
We'll probably need to fix this description, but I'll address this when reviewing this driver.

FRASTM reviewed (commented):
The stm32_clock_control_init function return code seems not tested yet, however if it could be <>0 , then we could see in the function something like 
```
if (IS_ENABLED(STM32_PLL_SRC_HSE)) {
...
if (IS_ENABLED(STM32_PLL3_P_ENABLE)) {
...
else
return -ENOTSUP
```

erwango committed:
drivers/clock_control: stm32u5: Use LL API for LSESYS programming

Use LL API when possible.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
DNM as this will likely require more work in short term to support multiple/alternate peripheral source clocks.

erwango committed:
drivers/clock_control/stm32h7: Independent clocks configuration

Rework clock start up functions in order to allow configuration
and enabling of individual clocks.
This way, each clock defined with a "okay" status will be enabled
even if not part of the sysclock clock tree.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -510,102 +506,117 @@ static struct clock_control_driver_api stm32_clock_control_api = {
+	}
+#endif /* STM32_PLL3_ENABLE */
 
-#if STM32_HSE_BYPASS
-	LL_RCC_HSE_EnableBypass();
-#else
-	LL_RCC_HSE_DisableBypass();
-#endif
```
This is now missing

erwango replied:
Has now been fixed

erwango committed:
include/drivers/clock_control: stm32: Add XXX_ENABLED and XXX_FREQ

Add STM32_FOO_ENABLED and STM32_FOO_FREQ to STM32 fixed clocks:
HSI, HSE, MSI(S), CSI, LSI, LSE..

Replace STM32_LSE_CLOCK by STM32_LSE_FREQ and when possible
replace by new STM32_LSE_ENABLED when making sense.

Fix STM32_PLL3_FOO_ENABLE to STM32_PLL3_FOO_ENABLED

Additionally, add STM32_PLL_FOO_ENABLED definitions.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/clock_control: stm32h7: Simplify SysClk freq computing

Instead of reading registers query the info on sysclock configuration
from existing configuration symbols.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

ABOSTM commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -501,163 +501,163 @@ static struct clock_control_driver_api stm32_clock_control_api = {
-		LL_RCC_PLL_SetSource(LL_RCC_PLLSOURCE_HSI);
-	} else {
-		return -ENOTSUP;
+	if (IS_ENABLED(STM32_PLL_ENABLED) || IS_ENABLED(STM32_PLL3_ENABLED)) {
```
`if (IS_ENABLED(STM32_PLL_ENABLED) || IS_ENABLED(STM32_PLL3_ENABLED)) {`
Is redundant with 
`#if defined(STM32_PLL_ENABLED) || defined(STM32_PLL3_ENABLED)`
few lines above

erwango replied:
Good catch, removed, thanks

FRASTM reviewed (dismissed):
change commit msg to drivers/clock_control: stm32u5: Use LL API for L**S**ESYS programming

erwango commented:
> change commit msg to drivers/clock_control: stm32u5: Use LL API for L**S**ESYS programming

Fixed, Thanks
